Address final review comments

shin19991207 opened the pull request:
Address #15 and #16

shin19991207 committed:
Update README.adoc

address https://github.com/OpenLiberty/draft-guide-microprofile-config-apis/issues/16

shin19991207 committed:
Update content and structure of the guide

shin19991207 committed:
Fix hotspots

shin19991207 committed:
Update title

shin19991207 committed:
Update title

dmuelle commented on README.adoc:
```diff
@@ -34,13 +34,15 @@ Learn how to use MicroProfile Config APIs to access configuration details and ag
+You'll learn how to improve flexibility and maintainability of your application's configurations by using https://openliberty.io/docs/latest/reference/feature/mpConfig-3.0.html[MicroProfile Config^]. 
 
-This guide assumes you are familiar with https://openliberty.io/docs/latest/external-configuration.html[External configuration of microservices^] and https://openliberty.io/docs/latest/reference/feature/mpConfig-3.0.html[MicroProfile Config^]. If you're new to MicroProfile Config, you might want to start with the https://openliberty.io/guides/microprofile-config-intro.html[Separating configuratio ...
+MicroProfile Config streamlines the configuration process for microservices by providing a flexible mechanism for application configuration. Application can use the https://openliberty.io/docs/latest/reference/javadoc/microprofile-5.0-javadoc.html?package=org/eclipse/microprofile/config/package-frame.html&class=org/eclipse/microprofile/config/package-summary.html[MicroProfile Config API^] to elim ...
```
```suggestion
MicroProfile Config streamlines the configuration process for microservices by providing a flexible mechanism for application configuration. Applications can use the https://openliberty.io/docs/latest/reference/javadoc/microprofile-5.0-javadoc.html?package=org/eclipse/microprofile/config/package-frame.html&class=org/eclipse/microprofile/config/package-summary.html[MicroProfile Config API^] to eliminate the need for repetitive injection statements and simplify the handling of complex configurations with bulk-extraction of configuration properties, dynamic property expressions, and retrieval of multi-valued properties as lists. Additionally, the configuration source APIs allow for fine-grained control and access to the metadata and values of configuration properties, enabling greater flexibility and customization in managing application configurations.
```

dmuelle commented on README.adoc:
```diff
@@ -118,33 +120,35 @@ cd query
+The MicroProfile Config API is included in the MicroProfile dependency that is specified in your [hotspot file=0]`query/pom.xml` file. Look for the dependency with the [hotspot=mpconfig file=0]`microprofile` artifact ID. This dependency provides a library that allows you to use the MicroProfile Config API to externalize configurations for your microservices. The [hotspot=mpconfig file=1]`mpConfig ...
 
-* `${prop:default}` - Provides a `default` value after the `:` character if the expression doesn't find a value for the `prop` property.
+In MicroProfile Config, property expressions provide a way to set and expand variables in property values. To use property expressions, define a property with a `${expression}` syntax, where `expression` is the name of another configuration property. When the property is loaded, the `${expression}` syntax is replaced with the value of the referenced property.
```
i think this paragraph and the one at line 151 could be moved to the beginning of this section- that would set the context before the reader starts the config task

shin19991207 replied:
Resolved in https://github.com/OpenLiberty/draft-guide-microprofile-config-apis/pull/17/commits/3d195954e4f0bf9f6a80e3696735774b6102941d

dmuelle commented on README.adoc:
```diff
@@ -159,7 +163,11 @@ query/microprofile-config.properties
+
+The `${system.user}` and `${system.password}` expressions are replaced with their respective values when the [hotspot=userPassword file=2]`system.userPassword` property is loaded, resulting in `alice:alicepwd`.
+
+To use this configuration property,
```
```suggestion
To use this configuration property, add it to the `SystemResource` class.
```

dmuelle commented on README.adoc:
```diff
@@ -174,14 +182,25 @@ query/SystemResource.java
+== Injecting property as a list
 
-Configuration values are Strings. MicroProfile Config API has built-in converters that automatically convert configured Strings into target types and retrieve multi-valued properties as lists.
+Configuration values are purely Strings. MicroProfile Config API has built-in converters that automatically convert configured Strings into target types. When injecting a multi-valued property, MicroProfile Config API assumes that the property value is a comma-separated list of items. It then splits the value into individual items and returns them as a List of Strings.
```
```suggestion
Configuration values are just Strings. MicroProfile Config API has built-in converters that automatically convert configured Strings into target types. When injecting a multi-valued property, MicroProfile Config API assumes that the property value is a comma-separated list of items. It then splits the value into individual items and returns them as a List of Strings.
```

dmuelle commented on README.adoc:
```diff
@@ -174,14 +182,25 @@ query/SystemResource.java
 
 // ===========================================================================================
-== Retrieving property as a list
+== Injecting property as a list
```
what is the benefit of doing this- or an example case where it would be useful?

shin19991207 replied:
Edited in https://github.com/OpenLiberty/draft-guide-microprofile-config-apis/pull/17/commits/324a3685c77af4f7dfe3051751832cbd805d968f

dmuelle commented on README.adoc:
```diff
@@ -213,18 +232,20 @@ query/SystemResource.java
 == Using configuration source APIs
 
-MicroProfile Config provides https://openliberty.io/docs/latest/reference/javadoc/microprofile-5.0-javadoc.html?package=org/eclipse/microprofile/config/package-frame.html&class=org/eclipse/microprofile/config/package-summary.html[configuration source APIs^] for retrieving details about configuration properties. You can use the `ConfigValue` API class to retrieve various useful information about a ...
+MicroProfile Config provides https://openliberty.io/docs/latest/reference/javadoc/microprofile-5.0-javadoc.html?package=org/eclipse/microprofile/config/package-frame.html&class=org/eclipse/microprofile/config/package-summary.html[configuration source APIs^] for obtaining detail information about your application's configuration. These APIs can be used to retrieve various useful information about  ...
```
same here- whats the benefit?

shin19991207 replied:
Edited in https://github.com/OpenLiberty/draft-guide-microprofile-config-apis/pull/17/commits/ae426b2edaa323381d31a623506df60e6220e304

dmuelle commented on README.adoc:
```diff
@@ -257,26 +279,28 @@ query/ConfigResource.java
+== Injecting configuration as a bulk through @ConfigProperties annotation
 
-Microprofile Config provides a way to aggregate configuration properties that start with the same prefix into a single property class through the `@ConfigProperties` annotation.
+MicroProfile Config provides a practical way to aggregate related configuration properties into a single CDI bean by using the `@ConfigProperties` annotation. This annotation is applied to a plain old Java object (POJO) that serves as a holder for configuration properties, making it easier to retrieve. By injecting the annotated POJO as a CDI bean into a separate POJO, a cleaner and more organize ...
```
```suggestion
MicroProfile Config provides a practical way to aggregate related configuration properties into a single CDI bean by using the `@ConfigProperties` annotation. This annotation is applied to a plain old Java object (POJO) that serves as a holder for configuration properties, making it easier to retrieve. Injecting the annotated POJO as a CDI bean into a separate POJO provides a cleaner and more organized approach to managing configuration properties.
```

dmuelle reviewed (changes requested):
ID review- made a few suggestions- a couple sections could still do with a little more context, making the value of the proposed config more clear

shin19991207 committed:
Update README.adoc

Co-authored-by: David Mueller <48686014+dmuelle@users.noreply.github.com>

shin19991207 committed:
Update README.adoc

Co-authored-by: David Mueller <48686014+dmuelle@users.noreply.github.com>

shin19991207 committed:
Update README.adoc

Co-authored-by: David Mueller <48686014+dmuelle@users.noreply.github.com>

shin19991207 committed:
Update README.adoc

Co-authored-by: David Mueller <48686014+dmuelle@users.noreply.github.com>

shin19991207 committed:
Update README.adoc

move line 149 and 151 to the beginning of the section

shin19991207 committed:
Update README.adoc

Specify the benefits and an example case where it would be useful of injecting property as a list.

shin19991207 committed:
Update README.adoc

Add the benefits of using config source APIs

dmuelle commented on README.adoc:
```diff
@@ -213,18 +234,22 @@ query/SystemResource.java
 == Using configuration source APIs
 
-MicroProfile Config provides https://openliberty.io/docs/latest/reference/javadoc/microprofile-5.0-javadoc.html?package=org/eclipse/microprofile/config/package-frame.html&class=org/eclipse/microprofile/config/package-summary.html[configuration source APIs^] for retrieving details about configuration properties. You can use the `ConfigValue` API class to retrieve various useful information about a ...
+MicroProfile Config provides https://openliberty.io/docs/latest/reference/javadoc/microprofile-5.0-javadoc.html?package=org/eclipse/microprofile/config/package-frame.html&class=org/eclipse/microprofile/config/package-summary.html[configuration source APIs^] for obtaining detail information about your application's configuration. These APIs can be used to retrieve various useful information about  ...
```
```suggestion
MicroProfile Config provides https://openliberty.io/docs/latest/reference/javadoc/microprofile-5.0-javadoc.html?package=org/eclipse/microprofile/config/package-frame.html&class=org/eclipse/microprofile/config/package-summary.html[configuration source APIs^] for obtaining detailed information about your application configuration. These APIs can be used to retrieve various useful information about a single specified configuration property by using the `ConfigValue` API class, or to look up the metadata of all underlying configurations of your microservice by using the `Config` API class.
```

dmuelle commented on README.adoc:
```diff
@@ -213,18 +234,22 @@ query/SystemResource.java
-MicroProfile Config provides https://openliberty.io/docs/latest/reference/javadoc/microprofile-5.0-javadoc.html?package=org/eclipse/microprofile/config/package-frame.html&class=org/eclipse/microprofile/config/package-summary.html[configuration source APIs^] for retrieving details about configuration properties. You can use the `ConfigValue` API class to retrieve various useful information about a ...
+MicroProfile Config provides https://openliberty.io/docs/latest/reference/javadoc/microprofile-5.0-javadoc.html?package=org/eclipse/microprofile/config/package-frame.html&class=org/eclipse/microprofile/config/package-summary.html[configuration source APIs^] for obtaining detail information about your application's configuration. These APIs can be used to retrieve various useful information about  ...
+
+The configuration source APIs offer a powerful toolset for managing and utilizing your microservice's configuration in a flexible and efficient manner. They enable you to add configurations from various sources, such as the `server.xml`, `server.env`, and `bootstrap.properties` files, or ConfigSources for different environments, such as `development`, `testing`, and `production`. With these APIs, ...
```
```suggestion
The configuration source APIs offer a powerful toolset for managing and utilizing your microservice configuration in a flexible and efficient manner. They enable you to add configurations from various sources, such as the `server.xml`, `server.env`, and `bootstrap.properties` files, or ConfigSources for different environments, such as `development`, `testing`, and `production`. With these APIs, you can easily determine which sources and properties are loaded in your current environment and make changes accordingly. These APIs can also be useful for debugging and troubleshooting purposes.
```

dmuelle commented on README.adoc:
```diff
@@ -97,6 +99,10 @@ mvn liberty:stop
 // =================================================================================================
 == Defining a property with expressions
 
+In MicroProfile Config, property expressions provide a way to set and expand variables in property values. To use property expressions, define a property with a `${expression}` syntax, where `expression` is the name of another configuration property. When the property is loaded, the `${expression}` syntax is replaced with the value of the referenced property.
```
I think the value of property expressions is that you can make the value of one property depend on the value of another property- can you think of a concise example where this would be beneficial? perhaps when you need to specify a port value or similar? Doesnt need to be much but could go in between the first and second sentence in this paragraph.

shin19991207 replied:
Edited in https://github.com/OpenLiberty/draft-guide-microprofile-config-apis/pull/17/commits/cdee87d7a502e51b1c3328cfe2e7972aaf4d5a3c

dmuelle reviewed (changes requested):
Nice work on adding context! Made a few more minor suggestions and request for a little more context on property expressions

shin19991207 committed:
Update README.adoc

Co-authored-by: David Mueller <48686014+dmuelle@users.noreply.github.com>

shin19991207 committed:
Update README.adoc

Co-authored-by: David Mueller <48686014+dmuelle@users.noreply.github.com>

shin19991207 committed:
Update README.adoc

Add an example where property expression is beneficial

shin19991207 committed:
Merge branch 'final-review' of github.com:OpenLiberty/draft-guide-microprofile-config-apis into final-review

dmuelle reviewed (approved):
LGTM thanks!

shin19991207 committed:
Update title

shin19991207 committed:
Update README.adoc

add backtick for enter/return

shin19991207 committed:
Update README.adoc

Edited :page-description:
